File: 2017-12/2017-12-02/leetcode1135.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<Integer>();
        for(int i = 0; i < asteroids.length; i++)
        {
            if(stack.empty() || asteroids[i] > 0)
            {
                stack.push(asteroids[i]);
                continue;
            }
            
            while(true)
            {
                int pre = stack.peek();
                if(pre < 0)
                {
                    stack.push(asteroids[i]);
                    break;
                }
                else if(pre == -asteroids[i])
                {
                    stack.pop();
                    break;
                }
                else if(pre > -asteroids[i])
                {
                    break;
                }
                else   // pre < -asteroids[i]
                {
                    stack.pop();
                    if(stack.isEmpty()){  
                        stack.push(asteroids[i]);  
                        break;  
                    }  
                }
            }
        }
        int[] res = new int[stack.size()];
        int i = stack.size()-1;
        while(!stack.empty())
        {
            res[i--] = stack.pop();
        }
        return res;
    }
}

File: 2017-12/2017-12-02/leetcode1135.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> st;
        for(int i = 0; i < asteroids.size(); i++)
        {
            if(asteroids[i] > 0)
            {
                st.push(asteroids[i]);
            }
            else if(asteroids[i] < 0)
            {
                while(!st.empty() && st.top() > 0 && st.top() < -asteroids[i])
                {
                    st.pop();
                }
                if(st.empty() || st.top() < 0) st.push(asteroids[i]);
                else if(st.top() == -asteroids[i]) st.pop();
                
            }
        }
        vector<int> res;
        while(!st.empty())
        {
            res.push_back(st.top());
            st.pop();
        }
        reverse(res.begin(), res.end());
        return res;
    }
};